gax-java/gax/src/main/java/com/google/api/gax/rpc/RequestParamsBuilder.java

```diff
@@ -56,19 +59,49 @@ public static RequestParamsBuilder create() {
    * pre-configured path templates. This method is called repeatedly for each configured routing
    * rule parameter, it's possible that the incoming field value from request is null or there is no
    * matches found, we'll continue the match-and-extract process for the next routing rule parameter
-   * in such case.
+   * in such case. This method will percent encode both the header key and header value.
    *
    * @param fieldValue the field value from a request
    * @param headerKey the header key for the routing header param
    * @param pathTemplate {@link PathTemplate} the path template used for match-and-extract
    */
   public void add(String fieldValue, String headerKey, PathTemplate pathTemplate) {
-    if (fieldValue == null) {
+    if (fieldValue == null || fieldValue.isEmpty()) {
       return;
     }
     Map<String, String> matchedValues = pathTemplate.match(fieldValue);
     if (matchedValues != null && matchedValues.containsKey(headerKey)) {
-      paramsBuilder.put(headerKey, matchedValues.get(headerKey));
+      String encodedKey = percentEncodeString(headerKey);
+      String encodedValue = percentEncodeString(matchedValues.get(headerKey));
+      paramsBuilder.put(encodedKey, encodedValue);
+    }
+  }
+
+  /**
+   * Add an entry to paramsBuilder with key-value pairing of (headerKey, fieldValue). The only
+   * validation done is to ensure the headerKey and fieldValue are not null and non-empty. This
+   * method will percent encode both the header key and header value.
+   *
+   * @param headerKey the header key for the routing header param
+   * @param fieldValue the field value from a request
+   */
+  public void add(String headerKey, String fieldValue) {
+    if (headerKey == null || headerKey.isEmpty() || fieldValue == null || fieldValue.isEmpty()) {
+      return;
+    }
+    String encodedKey = percentEncodeString(headerKey);
+    String encodedValue = percentEncodeString(fieldValue);
+    paramsBuilder.put(encodedKey, encodedValue);
+  }
+
+  // Percent encode the value passed in
+  private String percentEncodeString(String value) {
```

Comment:
[RequestUrlParamsEncoder](https://github.com/googleapis/sdk-platform-java/blob/cc5724f04722e3d60eb9b4e02ecc1bdf5dd7a7fa/gax-java/gax/src/main/java/com/google/api/gax/rpc/RequestUrlParamsEncoder.java#L79) might be doing the same thing already.

Reply:
This was my worry: https://github.com/googleapis/sdk-platform-java/blob/cc5724f04722e3d60eb9b4e02ecc1bdf5dd7a7fa/gax-java/gax/src/main/java/com/google/api/gax/rpc/RequestUrlParamsEncoder.java#L72-L73

I don't think it is encoding the values

Reply:
I see, the name is really confusing, it is more like a `paramsConcatenater`. This [isValid](https://github.com/googleapis/sdk-platform-java/blob/cc5724f04722e3d60eb9b4e02ecc1bdf5dd7a7fa/gax-java/gax/src/main/java/com/google/api/gax/rpc/RequestUrlParamsEncoder.java#L110-L121) method seems to be doing some validation but `validateExtractedParameters` is always false.